File: 338. Counting Bits.py
class Solution(object):
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        res = [0,1,1,2,1,2,2,3]
        ind = 4
        while len(res) <= num:
            temp = res[ind:]
            ind = len(res)
            res += temp
            res += [x+1 for x in temp]
        return res[:num+1]

File: 338. Counting Bits.cpp
class Solution {
public:
    vector<int> countBits(int num) {
        vector <int> res {0,1,1,2,1,2,2,3};
        int ind = 4;
        while (res.size() <= num)
        {
            vector <int> temp (res.begin() + ind, res.end());
            ind = res.size();
            res.insert(res.end(), temp.begin(), temp.end());
            for (int i = 0; i < temp.size(); i++)
                temp[i]++;
            res.insert(res.end(), temp.begin(), temp.end());
        }
        vector <int> rres (res.begin(), res.begin() + num + 1);
        return rres;
    }
};
